src/lem/path.rs

```diff
@@ -0,0 +1,296 @@
+
+use super::{interpreter::Frame, store::Store, symbol::Symbol, tag::Tag, MetaPtr, LEMOP};
+
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub(crate) enum PathNode {
+    Tag(Tag),
+    Symbol(Symbol),
+    Default,
+}
+
+impl std::fmt::Display for PathNode {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Tag(tag) => write!(f, "Tag({})", tag),
+            Self::Symbol(symbol) => write!(f, "Symbol({})", symbol),
+            Self::Default => write!(f, "Default"),
+        }
+    }
+}
+
+#[derive(Default, Clone, PartialEq, Eq, Hash)]
+pub struct Path(Vec<PathNode>);
+
+impl std::fmt::Display for Path {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let strings = self.0.iter().map(|x| format!("{}", x)).collect::<Vec<_>>();
+        write!(f, "{}", strings.join("."))
+    }
+}
+
+impl Path {
+    pub fn push_tag(&self, tag: &Tag) -> Path {
+        let mut path = self.0.clone();
+        path.push(PathNode::Tag(*tag));
+        Path(path)
+    }
+
+    pub fn push_symbol(&self, symbol: &Symbol) -> Path {
+        let mut path = self.0.clone();
+        path.push(PathNode::Symbol(symbol.clone()));
+        Path(path)
+    }
+
+    pub fn push_default(&self) -> Path {
+        let mut path = self.0.clone();
+        path.push(PathNode::Default);
+        Path(path)
+    }
+
+    #[inline]
+    pub fn push_tag_inplace(&mut self, tag: &Tag) {
+        self.0.push(PathNode::Tag(*tag));
+    }
+
+    #[inline]
+    pub fn push_symbol_inplace(&mut self, symbol: &Symbol) {
+        self.0.push(PathNode::Symbol(symbol.clone()));
+    }
+
+    #[inline]
+    pub fn push_default_inplace(&mut self) {
+        self.0.push(PathNode::Default);
+    }
+}
+
+fn insert_many(
+    map: &mut HashMap<String, String>,
+    ptr: &[MetaPtr],
+    path: &Path,
+) -> Result<Vec<MetaPtr>> {
+    ptr.iter()
+        .map(|ptr| {
+            let new_name = format!("{}.{}", path, ptr.name());
+            if map.insert(ptr.name().clone(), new_name.clone()).is_some() {
+                bail!("{} already defined", ptr.name());
+            };
+            Ok(MetaPtr(new_name))
+        })
+        .collect::<Result<Vec<_>>>()
+}
+
+fn insert_one(map: &mut HashMap<String, String>, ptr: &MetaPtr, path: &Path) -> Result<MetaPtr> {
+    let new_name = format!("{}.{}", path, ptr.name());
+    if map.insert(ptr.name().clone(), new_name.clone()).is_some() {
+        bail!("{} already defined", ptr.name());
+    };
+    Ok(MetaPtr(new_name))
+}
+
+fn retrieve_many(map: &HashMap<String, String>, args: &[MetaPtr]) -> Result<Vec<MetaPtr>> {
+    args.iter()
+        .map(|ptr| {
+            let Some(src_path) = map.get(ptr.name()).cloned() else {
+                bail!("{} not defined", ptr.name());
+            };
+            Ok(MetaPtr(src_path))
+        })
+        .collect::<Result<Vec<_>>>()
+}
+
+fn retrieve_one(map: &HashMap<String, String>, ptr: &MetaPtr) -> Result<MetaPtr> {
```

Comment:
`retrieve_one`…`HashMap<String, String>`. This definition leaves the reader with very little insight into what this function is for. It would be great to use the type system, and perhaps naming (but especially the type system), to help. The existence of a `HashMap<String, String>` in `lurk-rs` is itself a worrisome sign.